File: SearchIn2DMatrix.cpp
// Time Complexity : o(n x m)
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : NO

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& mat, int target) {
        int n=mat.size();
        int m=mat[0].size();
        int row=0,col=m-1;
        while(row<n && col>=0) {
            if(mat[row][col]==target) {
                return true;
            } else if(mat[row][col]>target) {
                col--;
            } else {
                row++;
            }
        }
        
        return false;
    }
};

File: src/com/madhurima/SearchIn2DMatrix.java
/*
    Time Complexity = O(m+n)
    Space Complexity = O(1)
    Did this code successfully run on Leetcode : yes
 */

package com.madhurima;

public class SearchIn2DMatrix {
}

class SolutionB {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0){
            return false;
        }

        int m = matrix.length;
        int n = matrix[0].length;

        int row = 0;
        int col = n-1;

        while(row >= 0 && row < m  && col >= 0 && col < n){
            if(matrix[row][col] == target){
                return true;
            }else if(target < matrix[row][col]){
                col--;
            }else{
                row++;
            }
        }

        return false;

    }

    public static void main(String[] args) {
        SolutionB s  = new SolutionB();
        int[][] matrix = {{1,4,7,11,15},{2,5,8,12,19},{3,6,9,16,22},{10,13,14,17,24},{18,21,23,26,30}};
        int target = 5;
        System.out.println(s.searchMatrix(matrix,target));
    }
}
